LXC Refactor

terminalmage opened the pull request:
**This will need further work before it is ready to be merged. See below.**

This pull request refactors LXC support, bringing it up-to-date with Salt's
coding style guidelines, as well as adding several features/functions. It also
adds an initial draft of a tutorial to the docs.

There is one major change that I made which may be modified before this is
merged. I modified the way that profiles work (while leaving legacy behavior
backwards-compatible) so that they weren't all configured under a single key.
This allows a single profile to be overridden without requiring all profiles to
be re-defined. During a conversation with @thatch45 we discussed adding an
option to the config.get remote execution function that merges dictionaries, a
feature which would make it unnecessary to configure profiles under individual
top-level keys. We need to further discuss this before we move forward with
merging.

**To install this pull request for testing:**
1. Check out the pull request locally (see [steps 1-5 here](https://help.github.com/articles/checking-out-pull-requests-locally/) for instructions). Where the instructions reference `origin`, replace that with the name of the git remote you have created for upstream. If you don't have such a remote, run `git remote add upstream https://github.com/saltstack/salt.git`, and use `upstream` instead of `origin` when fetching in step 4.
2. Install using `setup.py install` or `pip install --editable=/path/to/git/checkout`.

terminalmage committed:
Properly set lxc.utsname and /etc/hostname

terminalmage committed:
Allow interface dict to be passed to salt.utils.network.ip_addrs{,6}

This will allow the lxc.info function leverage existing logic rather
than trying to parse ip/ifconfig output by itself.

terminalmage committed:
LXC re-factor (WIP)

terminalmage committed:
Lint fixes + fix bad config entry name

terminalmage committed:
Make arguments vertical for easier reading

terminalmage committed:
Correct incorrect info in docstring

terminalmage committed:
Add stdin and output_loglevel args to cmd.run_chroot

Also allow the cmd function to be overridden with a __context__
variable.

terminalmage committed:
Add missing gather_bootstrap_script function

This is required by lxc.bootsrap but was only present in the develop
branch.

terminalmage committed:
Stub for LXC tutorial

terminalmage committed:
Modify lxc.created state, remove lxc.set_pass state

This commit modifies the lxc.created state to catch exceptions raised by
the lxc execution module. Additional states will follow.

If passwords need to be set, they can be done with a module.run state.
This function is not stateful and thus would run every time, making it a
bad fit as a state.

terminalmage committed:
Further work on lxc refactor

Most functions are now refactored. Also added functions to match the
ones in cmdmod.py so that running commands within containers more
closely resembles running commands on minions themselves.

terminalmage committed:
Tweak section header

terminalmage committed:
Complete LXC refactor

terminalmage committed:
Add self.strerror to SaltException class

This helps avoid deprecation warnings that happen when self.message is
referenced.

terminalmage committed:
Fix creation of container from image

Also changed return data for create() to match destroy(), clone(), etc.

terminalmage committed:
Remove snapshot from clone args

Let it live in kwargs so that it can be part of a profile

terminalmage committed:
Merge lxc.cloned into lxc.created state, rename to lxc.present

Also remove another redundant state which can be run using module.run.

terminalmage committed:
Always show prior state even state didn't change

Also some further refactoring of the maintenance functions, including
adding a "restart" function and deprecating the "restart" argument in
lxc.start.

terminalmage committed:
Finish LXC state refactoring

terminalmage committed:
Unfreeze frozen container before stopping

Gracefully stopping a frozen container results in a significant delay,
at least in my testing. This commit unfreezes the container before
stopping it, if the container was frozen when the stop() function is
called.

terminalmage committed:
Don't use _marker for default kwarg match in create()

The kwargs are passed in as None by the lxc.present state if no value
was specified, so ignore the kwarg match and go for the profile match
unless a kwarg has a non-None value.

terminalmage committed:
LXC state fixes

These are things that needed to be fixed/adjusted as a result of issues
found in manual testing.

terminalmage committed:
Deprecate "orig" parameter in favor of "clone_from"

terminalmage committed:
Remove unsupported argument

terminalmage committed:
Add back "orig" param for lxc.clone

Also move bootstrap_delay to lxc.bootstrap

terminalmage committed:
Bugfixes from LXC refactor

Also reclassify the new functionality as being added for the Lithium
release

terminalmage committed:
Fixes after rebasing against develop

terminalmage committed:
Fix lxc.bootstrap

I inadvertently broke this when resolving the merge conflict as I was
rebasing, since the location where the retrieved bootstrap script was
modified between 2014.7 and develop.

terminalmage committed:
Update docstring/comments

terminalmage committed:
More bootstrap fixes

Also, don't raise exceptions when lxc.init fails, just report the errors
and let the init continue if possible.

terminalmage committed:
Clarify docstring

terminalmage committed:
Docstring and logging fixes

terminalmage committed:
Add lxc.templates function

terminalmage committed:
Add more info to LXC tutorial

terminalmage committed:
Add back lxc.started state for compatibility

terminalmage committed:
More info for LXC tutorial

terminalmage committed:
Fixup indentation

terminalmage committed:
Complete initial draft of LXC tutorial

This will need to be filled out, but this is a good starting point.

terminalmage committed:
Properly invoke lxc.cloned (compatibility) state

Also made some docstring fixes

terminalmage committed:
Fix docstring typo

terminalmage committed:
Replace references to "next feature release" with 2015.2.0

terminalmage committed:
Replace references to Lithium with 2015.2.0

terminalmage committed:
Replace references to Lithium with 2015.2.0

terminalmage committed:
Formatting tweak (for consistency)

terminalmage committed:
Add warnings about creating containers from rootfs image

terminalmage committed:
Restore legacy behavior to lxc.run_cmd

This restores the old behavior (accidentally removed) while still
showing the deprecation notice.

terminalmage committed:
Hide set_pass from lxc state docs

terminalmage committed:
Add compatibility state for lxc.set_pass

terminalmage committed:
Add compatibility state for lxc.set_pass

terminalmage committed:
No need to mention that Debian Jessie hasn't been released

terminalmage committed:
Add missing comma

terminalmage committed:
Update docstring for lxc.update_lxc_conf

terminalmage committed:
Add back lxc.edited_conf state, with deprecation notice

rallytime commented:
@terminalmage I was just starting to look at this, but it looks like it could use some rebasing and pylint love. I was going to grab the pylint myself and submit a PR against your branch, but there are errors in there that you should probably resolve like doubly defined functions and undefined variables. 

Let me know when you get a chance to fix that up and I'll take another look at testing this.

terminalmage commented:
Rebased and worked in python_shell changes that had been merged forward into 2015.2. Waiting on tests, will likely still need another commit to fix any remaining lint errors.

thatch45 commented:
Yes, got some lint left and some failing tests, almost there!

terminalmage committed:
Add missing import

kiorky commented:
still broken in many places. :)

kiorky commented:
(and not only on salt cloud compat)

terminalmage commented:
@kiorky yep, found one instance I didn't catch in the lint errors. As always, your input on the "many places" where it is broken is welcome.

kiorky commented:
And my statements for last monthes are not forgotten and still actives, sorry.

terminalmage committed:
Lint fixes

terminalmage commented:
LXC-related lint failures are now cleaned up, the one failure is unrelated: https://jenkins.saltstack.com/job/salt-pr-lint-n/1368/violations/file/tests/integration/__init__.py/

kiorky commented:
hahaha ... !

kiorky commented:
-not tested code :D -

terminalmage commented:
Moving on to integration tests now.

terminalmage commented:
OK, test failures seem to have been caused by a typo when I rebased yesterday. I fixed that and rebased a fixup into the commit where the breakage occurred.

terminalmage commented:
Had to rebase one more time, Thomas Jackson's runner event stuff needed to be merged into this PR.

terminalmage committed:
Ignore lack of CLI example in deprecated function

kiorky commented:
(and the changesets i quoted here are just about specific new breakings done in the recent changesets, a lot is already adressed in our fork)

terminalmage committed:
Fix unused import

terminalmage committed:
Disable LXC integration tests for now

terminalmage commented:
I've had to disable the integration tests temporarily as they are 1) not distro-agnostic enough, and 2) result in an ``lxc-destroy`` of ALL containers in the ``tearDown`` function, which is bad for those running the tests locally. I'll need to rewrite these.

terminalmage committed:
Move Pillar's merge_sources to salt.utils.dictupdate

terminalmage committed:
Update ref to salt.utils.dictupdate.merge()

terminalmage committed:
Add merging and custom delimiter to config.get

terminalmage committed:
Change profiles to use new config.get merging code

terminalmage committed:
Remove clone_from argument from lxc.clone

terminalmage committed:
Fix container size calculation

terminalmage committed:
Update the LXC tutorial

This updates the tutoral to reflect the new profile merging behavior.

thatch45 commented:
I think that the file hashing test failure is related

rallytime commented:
Interesting. I don't think I've seen either of those test failures happen lately. Also, I will be working on testing these new fixes today with my lxc setup.

terminalmage committed:
Fix missing import

terminalmage committed:
Fix seed marker removal command

I tried to "fix" this in an earlier commit when it was already working,
and ended up breaking it. Derp. This commit reverses that breakage.

terminalmage committed:
Remove "ubuntu" from default users for lxc.init on ubuntu hosts

Just because the host is ubuntu doesn't mean that it will be
initializing only ubuntu containers (see: the "download" template).

terminalmage committed:
Fix for dict merging

This fixes a case where the function attempts to merge a dict into a
non-dict.

terminalmage committed:
lxc.init: Don't try to default to 1024MB mem limit

Memory limits should be explicit. Not only that, setting the limit like
this by default could, on hosts with less RAM, result in an unbootable
container.

Also replaced unnecessary usage of dict.pop with the marginally quicker
dict.get.

terminalmage committed:
Update lxc.init runner docstring to reflect change in mem limit behavior

terminalmage committed:
docstring updates

terminalmage committed:
Fix handling of lxc.get_parameter

Raise exception instead of returning a cryptic bool. Also, since this
function only queries a single parameter, there is no reason to return a
dict, just return the value.

terminalmage committed:
Add 'active' argument to lxc.ls

terminalmage committed:
Docstring tweaks

terminalmage committed:
Another fix for merging dicts into non-dicts

rallytime commented:
Woot! Down to only one failure.

terminalmage committed:
Default value for base exception class' "message" param

terminalmage committed:
Don't permit an lxc.create without needed params

This avoids a cryptic CLI error that ends up being returned when you try
an lxc-create with no template.

rallytime commented:
@thatch45 Once these test are finished running, this has my golden stamp of approval. All of the manual testing of this pull request passed very well!

thatch45 commented:
Thanks @rallytime !

rallytime commented:
WOOT!
